### DISCUSSION TITLE
How can I access ContextData in my middleware?

### DISCUSSION THREAD

**Discussion by sgabler** [ChilliCream/hotchocolate#5895] [Q&A] [answered]
I am using HC 12 and enabled the operation complexity feature like this:

```csharp
                .ModifyRequestOptions(o =>
                {
                    // See https://chillicream.com/docs/hotchocolate/v12/security/operation-complexity
                    o.Complexity.Enable = true;
                    o.Complexity.DefaultComplexity = 1;
                    o.Complexity.DefaultResolverComplexity = 5;
                    o.Complexity.ApplyDefaults = true;
                    o.Complexity.ContextDataKey = "QueryComplexity";
                })
```

This works nicely and will add an entry `"QueryComplexity"` in the `ContextData` dictionary.

I also have a middleware that tracks requests. A simplified version looks like this:

```csharp
        public RequestTrackerMiddleware(RequestDelegate next, ITransmitter<ApiTrackerEvent> eventDataTransmitter)
        {
            _next = next;
            _eventDataTransmitter = eventDataTransmitter;
        }

        public async Task InvokeAsync(HttpContext context)
        {
            try
            {
                await ExtractValuesAsync(context);
            }
            finally
            {
                context.Request.Body.Seek(0, SeekOrigin.Begin);
                await _next(context);
            }
        }
```

It extracts some values from the HttpContext and sends them somewhere else for storing.

Now the million dollar question: How can I access the value that is stored in the `ContextData` dictionary here in this middleware, so I can also store it?

What I tried so far:
- Create a request middleware that sets the complexity as a header => Doesn't work, because the headers arrive empty in the RequestTrackerMiddleware
- Try to see if I can create a middleware that uses IRequestContext instead, but that throws an exception, because the code doesn't allow it
  UseMiddlewareExtensions.cs
  ```csharp
              if (parameters.Length == 0 || parameters[0].ParameterType != typeof(HttpContext))
              {
                  throw new InvalidOperationException(Resources.FormatException_UseMiddlewareNoParameters(InvokeMethodName, InvokeAsyncMethodName, nameof(HttpContext)));
              }
  ```

Due to time constraints I can't update to HC13 now, so an idea that would work on HC12 would be much appreciated.

Thank you

**Comment by PascalSenn** [comment, marked as answer]
@sgabler  I guess i would just use a request middleware. 
```csharp
new ServiceCollection()
  .AddGraphQLServer()|
  ....
  .UseRequest< RequestTrackerMiddleware>() 
  .UseDefaultPipeline()
  ```
  ```csharp
        public RequestTrackerMiddleware(RequestDelegate next, ITransmitter<ApiTrackerEvent> eventDataTransmitter)
        {
            _next = next;
            _eventDataTransmitter = eventDataTransmitter;
        }

        public async ValueTask InvokeAsync(IRequestContext context)
        {
            // We execute the operation first. If you need you opteration to first execute you need to register the middlewars manually
            await _next(context);
            context.Request.Body.Seek(0, SeekOrigin.Begin);
            await ExtractValuesAsync(context);
          
        }
```

   We execute the operation first. If you need you opteration to first execute you need to register the middlewars manually (like in https://github.com/ChilliCream/graphql-platform/blob/main-version-12/src/HotChocolate/Core/src/Execution/DependencyInjection/RequestExecutorBuilderExtensions.UseRequest.cs#L184-L206) ****

**Reply by sgabler** [reply]
thanks, that works nicely!
